File: card-lang/src/calculator/parser/symbols/equal.rs
use std::collections::HashMap;

use crate::types::Parser::Parser;
use crate::types::Parser::LedListener;
use crate::types::Parser::SymbolAndToken;
use crate::types::ReusedStructs::Node;
use crate::types::ReusedStructs::NodeType;
use crate::types::ReusedStructs::Token;


pub struct EqualLed{}

impl LedListener for EqualLed {
  fn run(self, left: Node, parser: Parser) -> Node {
    let name = left.values.unwrap().get("name").unwrap();
    let node = Node {
      node_type: NodeType::AssignNode,
      values: Some(HashMap::from([
        ("name".to_string(), *name),
      ])),
      branches: Some(HashMap::from([
        ("value".to_string(), parser.expression(2)),
      ])),
      args: None
    };
    let node_type = left.node_type;
    match node_type {
      NodeType::CallNode => {
        if left.args.is_none() {
          panic!("there should be args when making a call");
        }
        let args = left.args.unwrap();
        for arg in args {
          match arg.node_type {
            NodeType::IdentifierNode => {}
            _ => {
              panic!("invalid argument name");
            }
          }
        }
        node.args = Some(args);
      }
      NodeType::IdentifierNode => {}
      _ => {
        panic!("Invalid node type");
      }
    }
    return node
  }
}

File: card-lang-typescript/src/calculator/parser/symbols/equal.ts
import { IdentifierNode, LangNode, LangNodeType } from "../../../types/ReusedStructs";
import { ParserController, LedListener } from "../../../types/Parser";

export class EqualLed implements LedListener {

  run(left: LangNode, parser: ParserController): LangNode {
    switch(left.node_type){
      case LangNodeType.CallNode: {
        let args: Array<IdentifierNode> = left.args.map((arg)=>{
          if (arg.node_type !== LangNodeType.IdentifierNode){
            throw new Error("Invalid argument name");
          }
          return arg
        });
        for(var i = 0; i < left.args.length; i++){
        }
        return {
          node_type: LangNodeType.FunctionNode,
          values: { name: left.values.name },
          args: args,
          branches: { right: parser.expression(2) }
        }
      }
      case LangNodeType.IdentifierNode: {
        return {
          node_type: LangNodeType.AssignNode,
          values: { name: left.values.name },
          branches: { right: parser.expression(2) }
        }
      }
      default: {
        throw new Error("Invalid Node Type")
      }
    }
  }
}

